add network roles info (#49501)
ansible/ansible: docs/docsite/rst/network/getting_started/network_roles.rst

      - name: configure hostname
        ios_config:
          lines: hostname {{ inventory_hostname }}

      - name: configure DNS
        ios_config:
          lines: ip name-server {{dns}}

Next, we'll run this playbook using the ``ansible-playbook`` command.  For brevity we used ``-l`` option to limit the playbook to only executing on the **rtr1** node.

.. code-block:: bash

   [user@ansible ~]$ ansible-playbook playbook.yml -l rtr1

   PLAY [configure cisco routers] *************************************************

   TASK [configure hostname] ******************************************************
   changed: [rtr1]

   TASK [configure DNS] ***********************************************************
   changed: [rtr1]

   PLAY RECAP *********************************************************************
   rtr1                       : ok=2    changed=2    unreachable=0    failed=0

This playbook configured the hostname and DNS servers.  Let's verify that configuration on the Cisco IOS XE **rtr1** router:

.. code-block:: bash

   rtr1#sh run | i name
   hostname rtr1
   ip name-server 8.8.8.8 8.8.4.4

Convert the playbook into a role
---------------------------------

> Now that we've seen a simple playbook in action, let's convert this into a reusable role. The Ansible Galaxy command line tool has the ability to create the standard framework for a role from scratch using the ``ansible-galaxy init`` command.  While not a requirement this can make it really easy to quickly start creating a role.

Reviewer: ```suggestion
Now that we've seen a simple playbook in action, let's convert this into a reusable role. You can create the directory structure manually, or you can use ``ansible-galaxy init`` to create the standard framework for a role.
```

.. code-block:: bash

   [user@ansible ~]$ ansible-galaxy init system-demo
   [user@ansible ~]$ cd system-demo/
   [user@ansible system-demo]$ tree
   .
   ├── defaults
   │   └── main.yml
   ├── files
   ├── handlers
   │   └── main.yml
   ├── meta
   │   └── main.yml
   ├── README.md
   ├── tasks
   │   └── main.yml
   ├── templates
   ├── tests
   │   ├── inventory
   │   └── test.yml
   └── vars
     └── main.yml

For this first demonstration we'll only use the **tasks** and **vars** directories.  The directory structure would look as follows:

.. code-block:: bash

   [user@ansible system-demo]$ tree
   .
   ├── tasks
   │   └── main.yml
   └── vars
       └── main.yml

> We'll now separate out the ``vars`` section and the ``tasks`` section from our original Ansible Playbook.  Move the two tasks into the ``tasks/main.yml`` file.  The file will look as follows:

Reviewer: ```suggestion
Next, move the content of the ``vars`` and ``tasks`` sections from our original Ansible Playbook into the role. First, move the two tasks into the ``tasks/main.yml`` file:
```

.. code-block:: bash

   [user@ansible system-demo]$ cat tasks/main.yml
   ---
   - name: configure hostname
     ios_config:
       lines: hostname {{ inventory_hostname }}

   - name: configure DNS
     ios_config:
       lines: ip name-server {{dns}}

> Move the vars into the ``vars/main.yml`` file.  The file will look as follows:

Reviewer: ```suggestion
Next, move the variables into the ``vars/main.yml`` file:
```

.. code-block:: bash

   [user@ansible system-demo]$ cat vars/main.yml
   ---
   dns: "8.8.8.8 8.8.4.4"

> Finally we'll modify the original Ansible Playbook to remove the keyword ``tasks`` and the keyword ``vars`` and then add the keyword ``roles``  with the name of the role, in this case ``system-demo``.  The playbook will now look like this:

Reviewer: ```suggestion
Finally, modify the original Ansible Playbook to remove the ``tasks`` and ``vars`` sections and add the keyword ``roles``  with the name of the role, in this case ``system-demo``.  You'll have this playbook:
```

.. _using_network_roles:

*************************
Use Ansible network roles
*************************

Roles are sets of Ansible defaults, files, tasks, templates, variables, and other Ansible components that work together. As you saw on :ref:`first_network_playbook`, moving from a command to a playbook makes it easy to run multiple tasks and repeat the same tasks in the same order. Moving from a playbook to a role makes it even easier to reuse and share your ordered tasks. You can look at :ref:`Ansible Galaxy <ansible_galaxy>`, which lets you share your roles and use others' roles, either directly or as inspiration.

.. contents::
   :local:

Understanding roles
===================

> So what exactly is a role, and why should you care? Ansible roles are basically playbooks broken up into a known file structure.  The goal for roles is to create robust and reusable functions so users don't have to recreate a playbook from scratch.  So for example, you don't have to write your own DNS playbook. Instead, you specify a DNS server, and an Ansible role will configure it for you.  Roles will make adopting Ansible Network Automation faster, easier and in a more agnostic fashion resulting in freeing you from the burden of building the playbook logic yourself.

Reviewer: I think there are two steps here. The first step is from a playbook to a role - moving to roles makes sharing, reading, and updating your Ansible workflow easier. Users can write their own roles . . . 

The second step is using [supported? approved? certified? official? pre-packaged? curated? I'm not sure what term we use for them . . . ] roles written by the Ansible Network team. Using these roles means you don't have to reinvent the wheel. These roles cover the most common network use cases, the things everyone needs to do for their network infrastructure. Instead of writing and maintaining your own `create_vlan` or `manage_DNS` playbooks or roles, you can concentrate on designing, codifying and maintaining the parser templates that describe your network topologies and inventory, and let Ansible's network roles do the work.

.. _using_network_roles:

*************************
Use Ansible network roles
*************************

> Roles are sets of Ansible defaults, files, tasks, templates, variables, and other Ansible components that work together. As you saw on :ref:`first_network_playbook`, moving from a command to a playbook makes it easy to run multiple tasks and repeat the same tasks in the same order. Moving from a playbook to a role makes it even easier to reuse and share your ordered tasks. You can look at :ref:`Ansible Galaxy <ansible_galaxy>`, which lets you share your roles and use others' roles, either directly or as inspiration.

Reviewer 2: + Playbooks?

Author: not sure what this comment is referring to?  Hard to  tell as it's an entire paragraph. If it's something simple, do you want to just use 'suggested edits' button above so I can see directly what you'd like changed?

To simplify your workflow even further, the Ansible Network team has written a series of roles for common network use cases. Using these roles means you don't have to reinvent the wheel. Instead of writing and maintaining your own ``create_vlan`` playbooks or roles, you can concentrate on designing, codifying and maintaining the parser templates that describe your network topologies and inventory, and let Ansible's network roles do the work. See the `network-related roles <https://galaxy.ansible.com/ansible-network>`_ on Ansible Galaxy.

A sample DNS playbook
---------------------

To demonstrate the concept of what a role is, the example below is a single YAML file containing a two-task playbook.  This Ansible Playbook configures the hostname on a Cisco IOS XE device, then it configures the DNS (domain name system) servers.

> In the output below you will see three Linux commands commonly used.  The **cat** command allows us to quickly look and view the contents of a YAML file, while the **tree** command allows us to look at the directory structure of the current working directory.  Finally the **cd** command allows us to change directories.

Reviewer 2: Do we need to do Linux 101 for people?

Author: removed

> .. code-block:: bash

Reviewer 2: Formatting this as `yaml` (and removing `cat ...`) may make it easier to read

.. _using_network_roles:

******************************************
Use Ansible network roles
******************************************

The Ansible Network team develops and supports a set of `network-related roles <https://galaxy.ansible.com/ansible-network>`_ on Galaxy. You can use these roles to jumpstart your network automation efforts. These roles are updated approximately every two weeks to give you access to the latest Ansible networking content.

These roles come in the following categories:

> * **User roles** - You use these roles, such as `config_manager <https://galaxy.ansible.com/ansible-network/config_manager>`_ and `cloud_vpn <https://galaxy.ansible.com/ansible-network/cloud_vpn>`_, directly in your playbooks. These roles are platform/provider agnostic, allowing you to use the same roles and playbooks across different network platforms or cloud providers.

Reviewer: I'd add a sentence at the beginning here to describe what user roles do - something like "User roles focus on tasks, such as managing configuration or ????." Then add the part about using these roles directly in playbooks. I'd also add an example (either here, or elsewhere with a link from here).

> * **Platform provider roles** - Network user roles depend on these provider roles to implement their functions. For example, the `config_manager <https://galaxy.ansible.com/ansible-network/config_manager>`_ role  uses the  `cisco_ios <https://galaxy.ansible.com/ansible-network/cisco_ios>`_ provider role to implement tasks on Cisco IOS network devices.

Reviewer: I'd describe this in a more provider-role-centric way. Maybe something like "Provider roles translate between the user roles and the various network OSs, each of which has a different API. Each provider role accepts input from any user role and translates it for a specific network OS. For example, the `cisco_ios` provider role implements tasks on Cisco IOS network devices." Because that provider role accepts input from any user role, right? not just from `config_manager`.

Author: afaik the provider roles only support the user roles they are written to support. For example, cisco_ios provider role supports both config_manager and cloud_vpn, but cisco_iosxr doesn't support either. But I like the rest of your suggestion, and I'll change 'any user role' to supported user roles.

* **Cloud provider and provisioner roles** - Similarly, cloud user roles depend on cloud provider and provisioner roles to implement cloud functions for specific cloud providers. For example, the `cloud_vpn <https://galaxy.ansible.com/ansible-network/cloud_vpn>`_ role depends on the `aws <https://galaxy.ansible.com/ansible-network/aws>`_ provider role to communicate with AWS.

.. tip::

>     You need to install at least one platform provider role for your network user roles, and set ``ansible_network_provider`` to that provider (for example, ``ansible_network_provider: ansible-network.cisco_ios``). Galaxy automatically installs any other dependencies listed in the role details on Galaxy.

Reviewer: Instead of a tip, I'd put an example/illustration here, to make this section more immediately useful. How do I install roles? How do I know which roles are "platform provider" roles?  How do I check if I already have a role installed? Is there a naming convention?

Author: This is where I don't want to repeat  what is already in the Galaxy documentation. I'm trying to focus just on 'hey we have roles...go see'.  ...with a couple of gotchas I hit along the way like the network provider bit.

Roles are fully documented with examples in Galaxy on the **Read Me** tab for each role.

Network roles release cycle
===========================

> The Ansible network team releases updates and new roles every two weeks. The role details on Galaxy lists the role versions available, and you can look in the GitHub repository to find the changelog file (for example, `CHANGELOG.rst <https://github.com/ansible-network/cisco_ios/blob/devel/CHANGELOG.rst>`_ ) that lists what has changed in each version of the role.

Reviewer: How do I update the role(s) I installed last month to get the latest version?
